### PR TITLE
GrandTheftAutoSanAndreas

### PR DESCRIPTION
**Pull request by davidar**
I think I might need to handle the LODs a bit more intelligently, as there's some glitches related to that (some z-fighting and funky mipmaps when zoomed out). It could probably do with some more optimisation too, as the size of the full map is pushing the draw call count.

image

### CODE REVIEW

File: src/GrandTheftAuto3/item.ts [hunk 1]
```diff
@@ -114,15 +125,44 @@ export interface ItemPlacement {
 }
 
 export function parseItemPlacement(text: string): ItemPlacement {
-    let instances = [] as ItemInstance[];
+    const instances = [] as ItemInstance[];
```
Comment (magcius) [inline src/GrandTheftAuto3/item.ts]: fyi I usually write `const instances: ItemInstance[] = [];`

File: src/GrandTheftAuto3/program.glsl [hunk 2]
```diff
@@ -11,6 +11,7 @@ layout(row_major, std140) uniform ub_SceneParams {
     vec4 u_SkyBotColor;
     vec4 u_WaterColor;
     vec4 u_WaterOrigin;
+    float u_Time;
```
Comment (magcius) [inline src/GrandTheftAuto3/program.glsl]: I'll fix this for you later, but you really shouldn't put floats in uniform blocks, as all fields require 16 byte alignment in std140 layout. I typically use a `vec4 u_Misc` field and then do `#define u_Time (u_Misc.x)`.

File: src/GrandTheftAuto3/render.ts [hunk 3]
```diff
@@ -387,10 +399,10 @@ export class DrawKey {
     constructor(obj: ObjectDefinition, public zone: string) {
         if (obj.flags & ObjectFlags.DRAW_LAST) {
             this.renderLayer = GfxRendererLayer.TRANSLUCENT;
-            this.modelName = obj.modelName;
+            //this.modelName = obj.modelName;
```
Comment (magcius) [inline src/GrandTheftAuto3/render.ts]: Reason for this?

**Comment by magcius** [comment]
Merged manually.

File: src/GrandTheftAuto3/render.ts (hunk 3 above)
Comment (davidar) [reply src/GrandTheftAuto3/render.ts]: I (temporarily) commented that out to reduce the draw call count for the full map, at the expense of possibly causing more transparency sorting issues. Not yet sure of the best way to solve it.

**Pull request closed without merging**